src/shell/server.cpp

```diff
@@ -625,6 +629,40 @@ json server::info(std::shared_ptr<module_info const> const & mod_info, pos_info
     return j;
 }
 
+json server::widget_event(std::shared_ptr<module_info const> const & mod_info, pos_info const & pos, json const & message) {
+    json j;
+    try {
+        parse_breaking_at_pos(mod_info->m_id, mod_info, pos);
+    } catch (break_at_pos_exception & e) {
+        auto opts = m_ios.get_options();
+        auto env = m_initial_env;
+        if (auto snap = get_closest_snapshot(mod_info, e.m_token_info.m_pos)) {
+            env = snap->m_snapshot_at_end->m_env;
+            opts = snap->m_snapshot_at_end->m_options;
+        }
+        update_widget(env, opts, m_ios, m_path, *mod_info, get_info_managers(m_lt), pos, e, j, message);
+    } catch (throwable & ex) {}
+
+    return j;
+}
+
+task<server::cmd_res> server::handle_widget_event(server::cmd_req const & req) {
+    cancel(m_bg_task_ctok);
+    m_bg_task_ctok = mk_cancellation_token();
+    // [todo] Need some help on how multithreading works. Should all events happen synchronously?
```

Comment:
The current implementation preempts widget events if the second one is issued before the first one is finished.  Maybe more problematically, it is also cancelled on info requestst, which happen on hovering.

Reply:
Yes this is bad but I'm not sure what to do to fix it. The only motivation behind this code as it stands now is that it mirrors server::info directly above it. There are two behaviours that I would consider good: 
1. once a widget_event occurs all later widget_events should be dropped until the render for the first event finishes. This would be simplest and safest to implement but it might lead to unresponsive UI if the user clicks or hovers many times. 
2. maintain a queue of widget_events and try to apply them all sequentially. If (say) a button disappears after the first event, the second event which clicks that button is dropped.

What is the behaviour of cancel()? Does it block until the given task reaches a cancellation point?

Reply:
`cancel` does not block, it sets an atomic boolean flag.  Lots of functions check this flag and raise an `interrupt_exception` if it is set.

Both option 1 and 2 sound good to me.